R/create_covariates.R

```diff
@@ -1,217 +1,189 @@
   }
 
 
-  #--- layers ---# have not updated this section yet
-
-  # layer_options <- c(
-  #   "sinksfilled", "sinkroute", "dem_preproc", "slope_aspect_curve",
-  #   "tcatchment", "tca", "scatchment", "twi", "channelsnetwork",
-  #   "overlandflow", "overlandflow2", "multiresflatness", "multiresflatness2",
-  #   "multiresflatness3", "tri", "convergence", "openness",
-  #   "dah", "tpi", "ridgevalley", "mrn", "flowaccumulation",
-  #   "slopelength", "flowaccumulation2", "flowaccumulation3",
-  #   "flowpathlength", "flowpathlength2", "flowpathlength3", "lsfactor",
-  #   "solarrad", "convexity", "vertdistance", "tci_low",
-  #   "swi", "windexp", "texture", "protection", "vrm",
-  #   "mbi", "mscale_tpi", "relposition", "slopecurvatures",
-  #   "steepestslope", "upslopearea"
-  # )
-  #
-  #
-
-  #--- layers ---# have not updated this section yet
+  #--- layers ---#
 
-  if(!is.character(layers)){
+  layer_options <- c(
+    "sinksfilled", "sinkroute", "dem_preproc", "slope_aspect_curve",
+    "tcatchment", "tca", "scatchment", "twi", "channelsnetwork",
+    "overlandflow", "overlandflow2", "multiresflatness", "multiresflatness2",
+    "multiresflatness3", "tri", "convergence", "openness",
+    "dah", "tpi", "ridgevalley", "mrn", "flowaccumulation",
+    "slopelength", "flowaccumulation2", "flowaccumulation3",
+    "flowpathlength", "flowpathlength2", "flowpathlength3", "lsfactor",
+    "solarrad", "convexity", "vertdistance", "tci_low",
+    "swi", "windexp", "texture", "protection", "vrm",
+    "mbi", "mscale_tpi", "relposition", "slopecurvatures",
+    "steepestslope", "upslopearea"
+  )
 
-    stop(paste0("`layers` must be type character."), call. = FALSE)
 
+  if (!inherits(layers, "character")) {
+    cli::cli_abort("{.var layers} must be a character string of the names of variables")
   }
 
   if (isTRUE(layers == "all")) {
     layers <- layer_options
   }
 
+  # check the layers are correct
+  errorl <- setdiff(layers, layer_options)
 
-  ####### Error handling -- unspecified layers ############
-  err.layers <- setdiff(layers, layer_options)
-
-  if (length(err.layers) == 1) {
-    stop(paste(err.layers, "is not a valid covariate" ))
+  if (length(errorl) > 0) {
+    cli::cli_abort("Layers contains covariates that are not a valid options, please review the following layer inputs: { errorl}")
   }
 
-  if (length(err.layers) > 1) {
-    print(err.layers)
-    stop("Specified covariates are not a valid options" )
+
+  # check which modules to run and appropriate dependants are also run in order
+  moddirs <- dplyr::filter(moddir, stringr::str_detect(moddir$module, paste(layers, collapse = "|")))
+  mods <- moddirs$module
+  deps <- setdiff(unique(c(moddirs$dep1, moddirs$dep2)), moddirs$module)
+  deps <- deps[!grepl("^\\s*$", deps)]
+  if (length(deps > 0)) {
+    mods <- c(moddir$module, deps)
   }
 
 
+  # order the modules to be run
+  layers_to_call <- mods[order(match(mods, moddir$module))]
```

Comment:
I got duplicates in some cases

```suggestion
  layers_to_call <- unique(mods[order(match(mods, moddir$module))])
```